src/pip/_internal/network/auth.py

```diff
@@ -250,8 +252,15 @@ def handle_401(self, resp, **kwargs):
 
         parsed = urllib_parse.urlparse(resp.url)
 
+        # Query the keyring for credentials:
+        username, password = self._get_new_credentials(resp.url,
+                                                       allow_netrc=False,
+                                                       allow_keyring=True)
+
         # Prompt the user for a new username and password
-        username, password, save = self._prompt_for_password(parsed.netloc)
+        save = False
+        if not username or not password:
```

Comment:
Will this have an impact on storing API key entries in keyring? Not sure how prevalent that is but it can appear as only one of these being set.

Reply:
For context on single-part login credentials, see #6796.

Reply:
In other words, this should probably be: `if not username` only, right?

Reply:
So I took a closer look. Previously we did:

1. Check up to 4 places (including keyring) for credentials based on the request URL (`_get_new_credentials`)
2. Make request
3. Get 401 and unconditionally prompt user for credentials with the response URL (`handle_401`)
4. Make followup request (`handle_401`)
5. Get 401 and warn user (`warn_on_401`)

Now we do:

1. Check up to 3 places (not keyring) for credentials based on the request URL (`_get_new_credentials`)
2. Make request
3. Get 401 and check up to 4 places for credentials (including keyring) based on the response URL (`handle_401`)
5. If that doesn't return something reasonable, then prompt the user
4. Make followup request (`handle_401`)
5. Get 401 and warn user (`warn_on_401`)

I think there's a few things to note with the changes we're making here:

1. The response URL may be different from the request URL in case of redirects, so the second time getting the credentials may behave differently
2. In the case where the response URL and request URL are the same, it may just try the same credentials again incorrectly
3. The response URL shouldn't contain credentials, so we don't have to worry about retrying with the same embedded credentials

It seems pretty complicated and to make headway we'd probably want to refactor some of this class so it doesn't get worse. If we disable keyring for pypi URLs maybe that's enough? Especially if this is a change people want to get in soon.